metals/src/main/scala/scala/meta/internal/bsp/BspConnector.scala

```diff
@@ -119,84 +136,183 @@ class BspConnector(
+              .exists(details =>
+                details.getName() == buildTool.executableName
+              ) =>
+          buildTool
+      }
+      .map { possible =>
+        possible.executableName -> Left(possible)
+      }
+      .toMap
+
+    // These are build servers that already have a .bsp entry plus bloop if
+    // it's an option.
+    val availableServers: Map[String, Either[
+      BuildTool with BuildServerProvider,
+      BspConnectionDetails
+    ]] = {
       if (bloopPresent || buildTools.loadSupported().nonEmpty)
         new BspConnectionDetails(
           BloopServers.name,
           ImmutableList.of(),
           userConfig().currentBloopVersion,
           "",
           ImmutableList.of()
-        ) :: found
-      else found
+        ) :: foundServers
+      else foundServers
+    }.map { details =>
+      details.getName() -> Right(details)
+    }.toMap
+
+    val allPossibleServers = possibleServers ++ availableServers
+
+    /**
+     * Handles showing the user what they need to know after an attempt to
+     * generate a bsp config has happened.
+     */
+    def handleGenerationStatus(
+        buildTool: BuildTool,
+        status: BspConfigGenerationStatus
+    ): Boolean = status match {
+      case BspConfigGenerationStatus.Generated =>
+        tables.buildServers.chooseServer(buildTool.executableName)
+        true
+      case Cancelled => false
+      case Failed(exit) =>
+        exit match {
+          case Left(exitCode) =>
+            scribe.error(
+              s"Creation of .bsp/${buildTool.executableName} failed with exit code: $exitCode"
+            )
+            client.showMessage(
+              Messages.BspProvider.genericUnableToCreateConfig
+            )
+          case Right(message) =>
+            client.showMessage(
+              Messages.BspProvider.unableToCreateConfigFromMessage(
+                message
+              )
+            )
+        }
+        false
     }
 
-    availableServers match {
+    def handleServerChoice(
+        possibleChoice: Option[String],
+        currentSelectedServer: Option[String]
+    ) = {
+      possibleChoice match {
+        case Some(choice) =>
+          allPossibleServers(choice) match {
+            case Left(buildTool) =>
+              buildTool
+                .generateBspConfig(
+                  workspace,
+                  args => bspConfigGenerator.runUnconditionally(buildTool, args)
+                )
+                .map(status => handleGenerationStatus(buildTool, status))
+            case Right(connectionDetails)
+                if connectionDetails.getName == BloopServers.name && currentSelectedServer
+                  .contains(
+                    BspConnector.BLOOP_SELECTED
+                  ) =>
+              Future.successful(false)
+            case Right(details) if details.getName == BloopServers.name =>
+              tables.buildServers.chooseServer(BspConnector.BLOOP_SELECTED)
```

Comment:
Just a question - why Bloop is handled specifically?
Why it doesn't fall in case where 
```scala
case *  if currentSelected.contains(name) => 
   tables.buildServers.chooseServer(details.getName)
   Future.successful(true)
case * =>  
  Future.successful(false)   
```

Btw. I agree with @tgodzik  that probably having a specific type might be a case here.
Looking at this function I think that having smth like bellow would reduce the amount of branches that should be handled.
```scala
case class BspChoice(serverName: String, switch: () => Future[Unit])
object BspChoice {
   def bloop: BspChoice = ....
   def fromBspDetails(details: BspDetails): BspChoice = ...
   def fromBuildTool(buildTool: BuildTool): BspChoise = ...
}
```

Reply:
> Just a question - why Bloop is handled specifically?

I think part of this is legacy due to the fact that we don't actually use BSP discovery with Bloop, but rather special case it through and through. In this specific situation (which is just carried over from how it was before) we do an extra check if it's bloop to see if there is already a `.bloop/` dir or not. Since we aren't creating a BSP entry for it, we instead have to call `createBloopAndConnect()`, which is actually `slowConnectToBuildServer()` whereas in the other cases we are using `quickConnectToBuildServer()`.

Part of me wants to think that this could be cleaned up some, but if we are going to attempt that I think we should fully tackle https://github.com/scalameta/metals/issues/2188. There is a bunch of connection stuff still in `MetalsLanguageServer`, but it's really intertwined 😅, but it would be a nice refactor/cleanup for the future.

As for the `BspChoice` type, I'll give that a try!